jupyterhub/templates/hub/_helpers-passwords.tpl

```diff
@@ -5,35 +5,43 @@
     proxy.secretToken       / hub.config.JupyterHub.proxy_auth_token
     hub.cookieSecret        / hub.config.JupyterHub.cookie_secret
     auth.state.cryptoKey    / hub.config.CryptKeeper.keys
+
+    Note that lookup logic returns falsy value when run with
+    `helm diff upgrade`, so it is a bit troublesome to test.
 */}}
 
 {{/*
     Returns given number of random Hex characters.
 
-    In practice, it generates up to 100 randAlphaNum strings
+    In practice, it generates up to 25 randAlphaNum strings
     that are filtered from non-hex characters and augmented
-    to the resulting string that is finally trimmed down.
+    to the resulting string that is finally trimmed down. We
+    do it multiple times as on average we filter out 4/15 of
+    the characters.
 */}}
 {{- define "jupyterhub.randHex" -}}
     {{- $result := "" }}
-    {{- range $i := until 100 }}
-        {{- if lt (len $result) . }}
-            {{- $rand_list := randAlphaNum . | splitList "" -}}
```

Comment:
Maybe we should be generating random numbers and using `printf "%x"` to get hex chars instead. Something like this:

```
{{- $result := "" }}
{{- range $i := until . }}
{{- $randHexChar := mod (randNumeric 4 | atoi) 16 | printf "%x" }} 
{{- $result = print $result $randHexChar }}
{{- end }}
{{- $result }}
```

This line:

```
{{- $randHexChar := mod (randNumeric 4 | atoi) 16 | printf "%x" }} 
```

always generates a single random hex character:

- `randNumeric 4 | atoi` generates a random number in [0, 10^4) (a range evenly divisble by 16, but even if off by one, that last partial interval offsetting randomness is only 1 part in 625)
- `mod N 16` maps to the range 0-15
- `printf "%x"` represents a single number 0-15 as a single hex character

It's a little wasteful to sample 10^4 to generate a number in 2^4, but since we can only generate random numbers in base 10, this works.

Reply:
Ah nice solution!

I considered using `%x` but ended up tunnel visioning away from it because of the issue you have estimated.

Given the complicated the path i ended up taking, as well as knowledge that `randAlphaNum 1` has issues even though perhaps `randAlphaNum <larger number>` doesn't, I figure it is a good compromise to go with this more aesthetically pleasing code that is still random enough.

Reply:
Oh, there isn't even an issue, but if there were, it would be small! :heart:

Reply:
Addressed by 482bbe4